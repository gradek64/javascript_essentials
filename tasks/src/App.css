@font-face {
  font-family: 'open-sans';
  font-style: normal;
  font-weight: normal;
  src: local('Open Sans Regular'), local('OpenSans-Regular'),
    url('./OpenSans-Regular.woff2') format('woff2');
}

body {
  background-color: #d2d2d2;
  justify-content: center;
  display: grid;
  font-family: 'open sans';
  font-weight: bold;
}
::placeholder {
  color: white;
}

.product-list {
  display: grid;
  grid-template-columns: repeat(1, 120px 302px);
  column-gap: 3%;
  row-gap: 3%;
  background-color: white;
  align-items: center;
  padding-top: 10px;
}

.product-list table {
  border-collapse: collapse;
  border: 1px solid #626a80;
}

.product-list thead,
.product-list tfoot {
  color: #ffffff;
}

.product-list tbody,
.product-list p {
  color: #ffffff;
}

.product-list th,
.product-list td {
  border: 1px solid #626a80;
  padding: 5px;
}

.product-list th {
  background-color: #eb008c;
}

.product-list th:first-child {
  width: 300px;
}

.product-list th:last-child {
  width: 100px;
}

.product-list tr:nth-child(2n) {
  background-color: #21242c;
}

.product-list tr:nth-child(2n + 1) {
  background-color: #2a2f3d;
}

.product-list td:last-child {
  text-align: right;
}
.product-list label,
.product-list input {
  box-sizing: border-box;
  margin-bottom: 5px;
  padding: 3px 8px;
  background-color: #eb008c;
  color: #ffffff;
  border: 1px solid #eb008c;
  border-radius: 4px;
  line-height: 24px;
  outline: none;
  width: 90%;
}

.product-list label {
  white-space: nowrap;
  border: none;
  padding: 3px;
  color: #eb008c;
  background-color: white;
}

/* scrolling tbody */
tbody {
  display: block;
  height: 250px;
  overflow: auto;
  background-image: url(./images/wowcherLogo.png);
  background-size: 10% auto;
}
thead,
tfoot,
tbody tr {
  display: table;
  width: 100%;
  table-layout: fixed;
}
tfoot tr th:first-child {
  text-align: left;
}
