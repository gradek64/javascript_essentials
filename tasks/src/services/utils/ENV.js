/* eslint-disable max-len */
export const ENV_HOST = {
  BASE_URL: 'http://localhost:3000',
};
