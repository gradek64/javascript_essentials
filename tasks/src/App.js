import React, { Component } from 'react';
//custom
import './App.css';
import apiService from './services/api-service';

//sorting
const sortASC = (data, sortOn) => {
  return data.sort(function (a, b) {
    const x = a[sortOn].toUpperCase();
    const y = b[sortOn].toUpperCase();
    if (x > y) {
      return 1;
    }
    if (x < y) {
      return -1;
    }
    return 0;
  });
};
const formatNumber = (number) =>
  new Intl.NumberFormat('en', { minimumFractionDigits: 2 }).format(number);

class App extends Component {
  state = {
    data: null,
    all: null,
    search: '',
  };

  componentDidMount = async () => {
    const branch1 = await apiService.get('/api/branch1.json');
    const branch2 = await apiService.get('/api/branch2.json');
    const branch3 = await apiService.get('/api/branch3.json');

    const mergeRequests = await Promise.all([branch1, branch2, branch3]);

    //1. merge all
    const allProducts = mergeRequests.reduce((all, { products }) => {
      const productsMerged = [...products];
      all.push(...productsMerged);
      return all;
    }, []);

    //2. calculate revenue and add revenue property
    const productsWithRevenue = allProducts.map((element) => {
      const { unitPrice, sold } = element;
      return { ...element, revenue: unitPrice * sold };
    });

    //3. find duplicate data and merge them
    const duplicates = {};
    productsWithRevenue.forEach(function (element) {
      const { name } = element;
      const trimStirng = name.trim();
      if (!duplicates[trimStirng]) duplicates[trimStirng] = [element];
      duplicates[trimStirng].push(element);
    });

    //4. reduce data from duplicates
    const dataReduced = Object.entries(duplicates).map(([key, entries]) => {
      const revenueTotal = entries.reduce((total, { revenue }) => {
        return total + revenue;
      }, 0);
      const soldTotal = entries.reduce((total, { sold }) => {
        return total + sold;
      }, 0);
      const [first] = entries;

      return { ...first, sold: soldTotal, revenue: revenueTotal };
    }, 0);

    //5. format data
    const formatedData = dataReduced.map((entry) => {
      const { revenue } = entry;
      const format = formatNumber(revenue);
      return { ...entry, revenue: format };
    });

    //6. sort data alphabetically done in render so it takes filtered ones as well
    this.setState({
      data: formatedData,
      all: formatedData,
    });
  };

  filterData = (e) => {
    const { data, all } = this.state;
    const { value } = e.target;
    this.setState({ search: value });
    //at least 2 charackters
    if (value.length >= 2) {
      const searchResults = data.filter((entry, i) => {
        const regex = new RegExp(`^${value}`, 'i');
        if (regex.test(entry.name)) return entry;
      });
      this.setState({ data: searchResults.length === 0 ? all : searchResults });
    } else {
      this.setState({ data: all });
    }
  };
  render() {
    const { data, search } = this.state;
    const totalRevenue =
      data &&
      data.reduce(
        (total, { revenue }) => total + Number(revenue.replace(/[,]/g, '')),
        0
      );

    return (
      <>
        {!data && (
          <div>
            <p>{'Loading...'}</p>
          </div>
        )}
        {data && (
          <div className="product-list">
            <label>Search Products</label>
            <input
              placeholder="...filter data"
              type="text"
              value={search}
              onChange={this.filterData}
            />
            <table className="tableData">
              <thead>
                <tr>
                  <th>Product</th>
                  <th>Revenue</th>
                </tr>
              </thead>
              <tbody>
                {sortASC(data, 'name').map(({ name, revenue }) => (
                  <tr key={name}>
                    <th>{name}</th>
                    <th>{revenue}</th>
                  </tr>
                ))}
              </tbody>
              <tfoot>
                <tr>
                  <th>Total</th>
                  <th>{formatNumber(totalRevenue)}</th>
                </tr>
              </tfoot>
            </table>
          </div>
        )}
      </>
    );
  }
}

export default App;
