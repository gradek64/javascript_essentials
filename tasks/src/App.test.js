import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { mount, shallow, configure } from 'enzyme';
import 'jest-enzyme';
import branch1 from '../public/api/branch1.json';
import branch2 from '../public/api/branch2.json';
import branch3 from '../public/api/branch3.json';
import App from './App';
import { ENV_HOST } from './services/utils/ENV';

configure({
  adapter: new Adapter(),
});

const responses = {
  [`${ENV_HOST.BASE_URL}/api/branch1.json`]: branch1,
  [`${ENV_HOST.BASE_URL}/api/branch2.json`]: branch2,
  [`${ENV_HOST.BASE_URL}/api/branch3.json`]: branch3,
};

global.fetch = (endpoint) => {
  return Promise.resolve({
    json: () => Promise.resolve(responses[endpoint]),
  });
};

// based on https://blog.pragmatists.com/genuine-guide-to-testing-react-redux-applications-6f3265c11f63
const flushAllPromises = () => new Promise((resolve) => setImmediate(resolve));

export const flushRequestsAndUpdate = async (enzymeWrapper) => {
  await flushAllPromises();
  enzymeWrapper.update();
};

describe('initial render', () => {
  it('renders without crashing', () => {
    mount(<App />);
  });
  it('should render and take a snapshot', () => {
    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });
  it('should render and take a snapshot with data', async () => {
    const component = shallow(<App />);
    await flushRequestsAndUpdate(component);
    expect(component).toMatchSnapshot();
  });
});
describe('handles loading data', () => {
  it('renders loading text initially', async () => {
    const app = mount(<App />);
    expect(app).toHaveText('Loading...');
  });

  it('renders a table after data load', async () => {
    const app = mount(<App />);
    expect(app).toHaveText('Loading...');
    await flushRequestsAndUpdate(app);
    expect(app.find('table')).toExist();
  });

  it('renders rows with product name as key', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);

    expect(app.find('table tbody tr').at(56).key()).toEqual('Hominy');
    expect(app.find('table tbody tr').at(73).key()).toEqual('Lychee');
  });
});

describe('data calculation', () => {
  it('calculates total revenue of all branches', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    expect(app.find('tfoot th:last-child').text()).toEqual('2,995,251.36');
  });

  it('filters and comes back with single product', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    const changeEvent = { target: { value: 'pears' } };
    app.find('input').simulate('change', changeEvent);
    expect(app.find('tfoot th:last-child').text()).toEqual('16,235.70');
  });
  it('filters and comes back with multiple products', async () => {
    const filteredData =
      '<tbody>' +
      '<tr><th>Coconuts</th><th>14,317.26</th>' +
      '</tr><tr><th>Collard Greens</th><th>28,314.60</th>' +
      '</tr><tr><th>Corn</th><th>14,001.36</th></tr>' +
      '</tbody>';
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    const changeEvent = { target: { value: 'co' } };
    app.find('input').simulate('change', changeEvent);

    expect(app.find('table tbody').html()).toEqual(filteredData);
    expect(app.find('tfoot th:last-child').text()).toEqual('56,633.22');
  });
  it('it wont start filtering if only one character is provided', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    const changeEvent = { target: { value: 'c' } };
    app.find('input').simulate('change', changeEvent);
    expect(app.find('tfoot th:last-child').text()).toEqual('2,995,251.36');
  });
  it('when filters comes back with empty it keeps all records ', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    const changeEvent = { target: { value: 'not existint' } };
    app.find('input').simulate('change', changeEvent);
    expect(app.find('tfoot th:last-child').text()).toEqual('2,995,251.36');
  });
});

describe('sort data alphabetically', () => {
  it('should sort all data', async () => {
    const expectedData = ['Alfalfa Sprouts', '17,714.01', 'Apple', '18,943.11'];
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);

    expectedData.forEach((th, index) => {
      const entry = app.find('table tbody tr th').at(index).text();
      expect(entry).toBe(th);
    });
  });
  it('should sort filtered data', async () => {
    const expectedData = [
      'Coconuts',
      '14,317.26',
      'Collard Greens',
      '28,314.60',
      'Corn',
      '14,001.36',
    ];
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    const changeEvent = { target: { value: 'co' } };
    app.find('input').simulate('change', changeEvent);

    expectedData.forEach((th, index) => {
      const entry = app.find('table tbody tr th').at(index).text();
      expect(entry).toBe(th);
    });
  });
});
