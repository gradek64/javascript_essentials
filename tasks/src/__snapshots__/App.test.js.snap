// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render should render and take a snapshot 1`] = `
<Fragment>
  <div>
    <p>
      Loading...
    </p>
  </div>
</Fragment>
`;

exports[`initial render should render and take a snapshot with data 1`] = `
<Fragment>
  <div
    className="product-list"
  >
    <label>
      Search Products
    </label>
    <input
      onChange={[Function]}
      placeholder="...filter data"
      type="text"
      value=""
    />
    <table
      className="tableData"
    >
      <thead>
        <tr>
          <th>
            Product
          </th>
          <th>
            Revenue
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="Alfalfa Sprouts"
        >
          <th>
            Alfalfa Sprouts
          </th>
          <th>
            17,714.01
          </th>
        </tr>
        <tr
          key="Apple"
        >
          <th>
            Apple
          </th>
          <th>
            18,943.11
          </th>
        </tr>
        <tr
          key="Apricot"
        >
          <th>
            Apricot
          </th>
          <th>
            25,294.95
          </th>
        </tr>
        <tr
          key="Artichoke"
        >
          <th>
            Artichoke
          </th>
          <th>
            8,031.10
          </th>
        </tr>
        <tr
          key="Asian Pear"
        >
          <th>
            Asian Pear
          </th>
          <th>
            16,607.43
          </th>
        </tr>
        <tr
          key="Asparagus"
        >
          <th>
            Asparagus
          </th>
          <th>
            29,048.69
          </th>
        </tr>
        <tr
          key="Atemoya"
        >
          <th>
            Atemoya
          </th>
          <th>
            16,244.28
          </th>
        </tr>
        <tr
          key="Avocado"
        >
          <th>
            Avocado
          </th>
          <th>
            11,745.00
          </th>
        </tr>
        <tr
          key="Bamboo Shoots"
        >
          <th>
            Bamboo Shoots
          </th>
          <th>
            16,334.64
          </th>
        </tr>
        <tr
          key="Banana"
        >
          <th>
            Banana
          </th>
          <th>
            16,368.40
          </th>
        </tr>
        <tr
          key="Bean Sprouts"
        >
          <th>
            Bean Sprouts
          </th>
          <th>
            25,457.76
          </th>
        </tr>
        <tr
          key="Beans"
        >
          <th>
            Beans
          </th>
          <th>
            31,269.69
          </th>
        </tr>
        <tr
          key="Beets"
        >
          <th>
            Beets
          </th>
          <th>
            41,115.60
          </th>
        </tr>
        <tr
          key="Belgian Endive"
        >
          <th>
            Belgian Endive
          </th>
          <th>
            18,278.24
          </th>
        </tr>
        <tr
          key="Bell Peppers"
        >
          <th>
            Bell Peppers
          </th>
          <th>
            38,653.92
          </th>
        </tr>
        <tr
          key="Bitter Melon"
        >
          <th>
            Bitter Melon
          </th>
          <th>
            30,027.81
          </th>
        </tr>
        <tr
          key="Blackberries"
        >
          <th>
            Blackberries
          </th>
          <th>
            42,338.40
          </th>
        </tr>
        <tr
          key="Blueberries"
        >
          <th>
            Blueberries
          </th>
          <th>
            15,302.95
          </th>
        </tr>
        <tr
          key="Bok Choy"
        >
          <th>
            Bok Choy
          </th>
          <th>
            32,725.63
          </th>
        </tr>
        <tr
          key="Boniato"
        >
          <th>
            Boniato
          </th>
          <th>
            13,787.62
          </th>
        </tr>
        <tr
          key="Boysenberries"
        >
          <th>
            Boysenberries
          </th>
          <th>
            16,783.20
          </th>
        </tr>
        <tr
          key="Broccoflower"
        >
          <th>
            Broccoflower
          </th>
          <th>
            44,056.32
          </th>
        </tr>
        <tr
          key="Broccoli"
        >
          <th>
            Broccoli
          </th>
          <th>
            30,177.40
          </th>
        </tr>
        <tr
          key="Brussels Sprouts"
        >
          <th>
            Brussels Sprouts
          </th>
          <th>
            6,405.12
          </th>
        </tr>
        <tr
          key="Cabbage"
        >
          <th>
            Cabbage
          </th>
          <th>
            15,287.40
          </th>
        </tr>
        <tr
          key="Cactus Pear"
        >
          <th>
            Cactus Pear
          </th>
          <th>
            29,973.60
          </th>
        </tr>
        <tr
          key="Cantaloupe"
        >
          <th>
            Cantaloupe
          </th>
          <th>
            18,676.35
          </th>
        </tr>
        <tr
          key="Carambola"
        >
          <th>
            Carambola
          </th>
          <th>
            12,248.26
          </th>
        </tr>
        <tr
          key="Carrots"
        >
          <th>
            Carrots
          </th>
          <th>
            44,011.20
          </th>
        </tr>
        <tr
          key="Casaba Melon"
        >
          <th>
            Casaba Melon
          </th>
          <th>
            9,414.13
          </th>
        </tr>
        <tr
          key="Cauliflower"
        >
          <th>
            Cauliflower
          </th>
          <th>
            16,222.98
          </th>
        </tr>
        <tr
          key="Celery"
        >
          <th>
            Celery
          </th>
          <th>
            14,738.64
          </th>
        </tr>
        <tr
          key="Chayote"
        >
          <th>
            Chayote
          </th>
          <th>
            26,606.56
          </th>
        </tr>
        <tr
          key="Cherimoya"
        >
          <th>
            Cherimoya
          </th>
          <th>
            20,986.35
          </th>
        </tr>
        <tr
          key="Cherries"
        >
          <th>
            Cherries
          </th>
          <th>
            10,516.80
          </th>
        </tr>
        <tr
          key="Coconuts"
        >
          <th>
            Coconuts
          </th>
          <th>
            14,317.26
          </th>
        </tr>
        <tr
          key="Collard Greens"
        >
          <th>
            Collard Greens
          </th>
          <th>
            28,314.60
          </th>
        </tr>
        <tr
          key="Corn"
        >
          <th>
            Corn
          </th>
          <th>
            14,001.36
          </th>
        </tr>
        <tr
          key="Cranberries"
        >
          <th>
            Cranberries
          </th>
          <th>
            39,946.41
          </th>
        </tr>
        <tr
          key="Cucumber"
        >
          <th>
            Cucumber
          </th>
          <th>
            16,370.31
          </th>
        </tr>
        <tr
          key="Dates"
        >
          <th>
            Dates
          </th>
          <th>
            28,756.40
          </th>
        </tr>
        <tr
          key="Dried Plums"
        >
          <th>
            Dried Plums
          </th>
          <th>
            14,463.10
          </th>
        </tr>
        <tr
          key="Eggplant"
        >
          <th>
            Eggplant
          </th>
          <th>
            14,976.90
          </th>
        </tr>
        <tr
          key="Endive"
        >
          <th>
            Endive
          </th>
          <th>
            25,872.36
          </th>
        </tr>
        <tr
          key="Escarole"
        >
          <th>
            Escarole
          </th>
          <th>
            10,094.41
          </th>
        </tr>
        <tr
          key="Feijoa"
        >
          <th>
            Feijoa
          </th>
          <th>
            17,975.99
          </th>
        </tr>
        <tr
          key="Fennel"
        >
          <th>
            Fennel
          </th>
          <th>
            19,237.71
          </th>
        </tr>
        <tr
          key="Figs"
        >
          <th>
            Figs
          </th>
          <th>
            18,453.12
          </th>
        </tr>
        <tr
          key="Garlic"
        >
          <th>
            Garlic
          </th>
          <th>
            27,601.08
          </th>
        </tr>
        <tr
          key="Gooseberries"
        >
          <th>
            Gooseberries
          </th>
          <th>
            4,282.48
          </th>
        </tr>
        <tr
          key="Grapefruit"
        >
          <th>
            Grapefruit
          </th>
          <th>
            28,873.12
          </th>
        </tr>
        <tr
          key="Grapes"
        >
          <th>
            Grapes
          </th>
          <th>
            40,013.08
          </th>
        </tr>
        <tr
          key="Green Beans"
        >
          <th>
            Green Beans
          </th>
          <th>
            18,819.76
          </th>
        </tr>
        <tr
          key="Green Onions"
        >
          <th>
            Green Onions
          </th>
          <th>
            21,782.61
          </th>
        </tr>
        <tr
          key="Greens"
        >
          <th>
            Greens
          </th>
          <th>
            14,757.15
          </th>
        </tr>
        <tr
          key="Guava"
        >
          <th>
            Guava
          </th>
          <th>
            32,284.35
          </th>
        </tr>
        <tr
          key="Hominy"
        >
          <th>
            Hominy
          </th>
          <th>
            16,609.36
          </th>
        </tr>
        <tr
          key="Honeydew Melon"
        >
          <th>
            Honeydew Melon
          </th>
          <th>
            17,187.24
          </th>
        </tr>
        <tr
          key="Horned Melon"
        >
          <th>
            Horned Melon
          </th>
          <th>
            10,940.71
          </th>
        </tr>
        <tr
          key="Iceberg Lettuce"
        >
          <th>
            Iceberg Lettuce
          </th>
          <th>
            27,772.62
          </th>
        </tr>
        <tr
          key="Jerusalem Artichoke"
        >
          <th>
            Jerusalem Artichoke
          </th>
          <th>
            44,225.16
          </th>
        </tr>
        <tr
          key="Jicama"
        >
          <th>
            Jicama
          </th>
          <th>
            20,538.00
          </th>
        </tr>
        <tr
          key="Kale"
        >
          <th>
            Kale
          </th>
          <th>
            13,611.20
          </th>
        </tr>
        <tr
          key="Kiwifruit"
        >
          <th>
            Kiwifruit
          </th>
          <th>
            43,978.60
          </th>
        </tr>
        <tr
          key="Kohlrabi"
        >
          <th>
            Kohlrabi
          </th>
          <th>
            34,879.68
          </th>
        </tr>
        <tr
          key="Kumquat"
        >
          <th>
            Kumquat
          </th>
          <th>
            22,500.24
          </th>
        </tr>
        <tr
          key="Leeks"
        >
          <th>
            Leeks
          </th>
          <th>
            9,357.14
          </th>
        </tr>
        <tr
          key="Lemons"
        >
          <th>
            Lemons
          </th>
          <th>
            24,879.30
          </th>
        </tr>
        <tr
          key="Lettuce"
        >
          <th>
            Lettuce
          </th>
          <th>
            16,811.52
          </th>
        </tr>
        <tr
          key="Lima Beans"
        >
          <th>
            Lima Beans
          </th>
          <th>
            6,087.03
          </th>
        </tr>
        <tr
          key="Limes"
        >
          <th>
            Limes
          </th>
          <th>
            23,076.90
          </th>
        </tr>
        <tr
          key="Longan"
        >
          <th>
            Longan
          </th>
          <th>
            21,610.58
          </th>
        </tr>
        <tr
          key="Loquat"
        >
          <th>
            Loquat
          </th>
          <th>
            8,236.11
          </th>
        </tr>
        <tr
          key="Lychee"
        >
          <th>
            Lychee
          </th>
          <th>
            32,632.46
          </th>
        </tr>
        <tr
          key="Madarins"
        >
          <th>
            Madarins
          </th>
          <th>
            8,321.12
          </th>
        </tr>
        <tr
          key="Malanga"
        >
          <th>
            Malanga
          </th>
          <th>
            31,262.96
          </th>
        </tr>
        <tr
          key="Mandarin Oranges"
        >
          <th>
            Mandarin Oranges
          </th>
          <th>
            16,013.20
          </th>
        </tr>
        <tr
          key="Mangos"
        >
          <th>
            Mangos
          </th>
          <th>
            15,769.12
          </th>
        </tr>
        <tr
          key="Mulberries"
        >
          <th>
            Mulberries
          </th>
          <th>
            11,836.80
          </th>
        </tr>
        <tr
          key="Mushrooms"
        >
          <th>
            Mushrooms
          </th>
          <th>
            15,521.92
          </th>
        </tr>
        <tr
          key="Napa"
        >
          <th>
            Napa
          </th>
          <th>
            22,825.66
          </th>
        </tr>
        <tr
          key="Nectarines"
        >
          <th>
            Nectarines
          </th>
          <th>
            43,625.68
          </th>
        </tr>
        <tr
          key="Okra"
        >
          <th>
            Okra
          </th>
          <th>
            26,441.61
          </th>
        </tr>
        <tr
          key="Onion"
        >
          <th>
            Onion
          </th>
          <th>
            19,489.52
          </th>
        </tr>
        <tr
          key="Oranges"
        >
          <th>
            Oranges
          </th>
          <th>
            15,840.72
          </th>
        </tr>
        <tr
          key="Papayas"
        >
          <th>
            Papayas
          </th>
          <th>
            26,074.65
          </th>
        </tr>
        <tr
          key="Parsnip"
        >
          <th>
            Parsnip
          </th>
          <th>
            21,161.62
          </th>
        </tr>
        <tr
          key="Passion Fruit"
        >
          <th>
            Passion Fruit
          </th>
          <th>
            34,337.16
          </th>
        </tr>
        <tr
          key="Peaches"
        >
          <th>
            Peaches
          </th>
          <th>
            32,012.64
          </th>
        </tr>
        <tr
          key="Pears"
        >
          <th>
            Pears
          </th>
          <th>
            16,235.70
          </th>
        </tr>
        <tr
          key="Peas"
        >
          <th>
            Peas
          </th>
          <th>
            23,949.60
          </th>
        </tr>
        <tr
          key="Peppers"
        >
          <th>
            Peppers
          </th>
          <th>
            11,151.79
          </th>
        </tr>
        <tr
          key="Persimmons"
        >
          <th>
            Persimmons
          </th>
          <th>
            35,846.54
          </th>
        </tr>
        <tr
          key="Pineapple"
        >
          <th>
            Pineapple
          </th>
          <th>
            33,411.36
          </th>
        </tr>
        <tr
          key="Plantains"
        >
          <th>
            Plantains
          </th>
          <th>
            13,449.04
          </th>
        </tr>
        <tr
          key="Plums"
        >
          <th>
            Plums
          </th>
          <th>
            17,713.80
          </th>
        </tr>
        <tr
          key="Pomegranate"
        >
          <th>
            Pomegranate
          </th>
          <th>
            19,097.47
          </th>
        </tr>
        <tr
          key="Potatoes"
        >
          <th>
            Potatoes
          </th>
          <th>
            14,082.09
          </th>
        </tr>
        <tr
          key="Prickly Pear"
        >
          <th>
            Prickly Pear
          </th>
          <th>
            26,334.95
          </th>
        </tr>
        <tr
          key="Prunes"
        >
          <th>
            Prunes
          </th>
          <th>
            22,858.29
          </th>
        </tr>
        <tr
          key="Pummelo"
        >
          <th>
            Pummelo
          </th>
          <th>
            25,575.54
          </th>
        </tr>
        <tr
          key="Pumpkin"
        >
          <th>
            Pumpkin
          </th>
          <th>
            17,008.24
          </th>
        </tr>
        <tr
          key="Quince"
        >
          <th>
            Quince
          </th>
          <th>
            19,787.26
          </th>
        </tr>
        <tr
          key="Radicchio"
        >
          <th>
            Radicchio
          </th>
          <th>
            32,043.84
          </th>
        </tr>
        <tr
          key="Radishes"
        >
          <th>
            Radishes
          </th>
          <th>
            34,988.70
          </th>
        </tr>
        <tr
          key="Raisins"
        >
          <th>
            Raisins
          </th>
          <th>
            11,271.69
          </th>
        </tr>
        <tr
          key="Raspberries"
        >
          <th>
            Raspberries
          </th>
          <th>
            7,023.61
          </th>
        </tr>
        <tr
          key="Red Cabbage"
        >
          <th>
            Red Cabbage
          </th>
          <th>
            15,204.80
          </th>
        </tr>
        <tr
          key="Rhubarb"
        >
          <th>
            Rhubarb
          </th>
          <th>
            25,516.40
          </th>
        </tr>
        <tr
          key="Romaine Lettuce"
        >
          <th>
            Romaine Lettuce
          </th>
          <th>
            37,937.13
          </th>
        </tr>
        <tr
          key="Rutabaga"
        >
          <th>
            Rutabaga
          </th>
          <th>
            25,961.44
          </th>
        </tr>
        <tr
          key="Shallots"
        >
          <th>
            Shallots
          </th>
          <th>
            7,994.94
          </th>
        </tr>
        <tr
          key="Snow Peas"
        >
          <th>
            Snow Peas
          </th>
          <th>
            16,122.66
          </th>
        </tr>
        <tr
          key="Spinach"
        >
          <th>
            Spinach
          </th>
          <th>
            25,696.98
          </th>
        </tr>
        <tr
          key="Sprouts"
        >
          <th>
            Sprouts
          </th>
          <th>
            37,881.00
          </th>
        </tr>
        <tr
          key="Squash"
        >
          <th>
            Squash
          </th>
          <th>
            24,640.00
          </th>
        </tr>
        <tr
          key="Strawberries"
        >
          <th>
            Strawberries
          </th>
          <th>
            14,737.24
          </th>
        </tr>
        <tr
          key="String Beans"
        >
          <th>
            String Beans
          </th>
          <th>
            31,501.94
          </th>
        </tr>
        <tr
          key="Sweet Potato"
        >
          <th>
            Sweet Potato
          </th>
          <th>
            42,684.00
          </th>
        </tr>
        <tr
          key="Tangelo"
        >
          <th>
            Tangelo
          </th>
          <th>
            30,480.45
          </th>
        </tr>
        <tr
          key="Tangerines"
        >
          <th>
            Tangerines
          </th>
          <th>
            37,172.30
          </th>
        </tr>
        <tr
          key="Tomatillo"
        >
          <th>
            Tomatillo
          </th>
          <th>
            15,888.60
          </th>
        </tr>
        <tr
          key="Tomato"
        >
          <th>
            Tomato
          </th>
          <th>
            23,331.24
          </th>
        </tr>
        <tr
          key="Turnip"
        >
          <th>
            Turnip
          </th>
          <th>
            21,887.58
          </th>
        </tr>
        <tr
          key="Ugli Fruit"
        >
          <th>
            Ugli Fruit
          </th>
          <th>
            16,515.20
          </th>
        </tr>
        <tr
          key="Water Chestnuts"
        >
          <th>
            Water Chestnuts
          </th>
          <th>
            12,710.31
          </th>
        </tr>
        <tr
          key="Watercress"
        >
          <th>
            Watercress
          </th>
          <th>
            16,234.65
          </th>
        </tr>
        <tr
          key="Watermelon"
        >
          <th>
            Watermelon
          </th>
          <th>
            26,786.76
          </th>
        </tr>
        <tr
          key="Waxed Beans"
        >
          <th>
            Waxed Beans
          </th>
          <th>
            14,843.07
          </th>
        </tr>
        <tr
          key="Yams"
        >
          <th>
            Yams
          </th>
          <th>
            15,865.71
          </th>
        </tr>
        <tr
          key="Yellow Squash"
        >
          <th>
            Yellow Squash
          </th>
          <th>
            27,249.46
          </th>
        </tr>
        <tr
          key="Yuca/Cassava"
        >
          <th>
            Yuca/Cassava
          </th>
          <th>
            60,554.50
          </th>
        </tr>
        <tr
          key="Zucchini Squash"
        >
          <th>
            Zucchini Squash
          </th>
          <th>
            19,968.16
          </th>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <th>
            Total
          </th>
          <th>
            2,995,251.36
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
</Fragment>
`;
